File: common/src/main/java/su/plo/voice/event/VoiceEventBus.java
package su.plo.voice.event;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Maps;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import su.plo.voice.addon.VoiceAddonManager;
import su.plo.voice.api.PlasmoVoice;
import su.plo.voice.api.addon.annotation.Addon;
import su.plo.voice.api.event.*;

import java.lang.reflect.Method;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public final class VoiceEventBus implements EventBus {

    private static final Logger LOGGER = LogManager.getLogger(VoiceAddonManager.class);

    // listener -> event handlers
    private final Map<Object, List<EventHandler<?>>> registeredListeners = Maps.newConcurrentMap();

    // addon -> listeners
    private final Map<Object, List<Object>> registeredAddonListeners = Maps.newConcurrentMap();

    // addon -> event handlers
    private final Map<Object, List<EventHandler<?>>> registeredAddonHandlers = Maps.newConcurrentMap();

    // class -> handlers
    private final Map<Class<? extends Event>, EnumMap<EventPriority, List<EventHandler<?>>>> handlers = Maps.newConcurrentMap();

    private final Executor asyncExecutor = Executors.newSingleThreadExecutor();

    @Override
    public <E extends Event> void call(@NotNull E event) {
        if (!this.handlers.containsKey(event.getClass())) return;

        for (Map.Entry<EventPriority, List<EventHandler<?>>> entry :
                this.handlers.get(event.getClass()).entrySet()) {
            List<EventHandler<?>> listeners = entry.getValue();

            for (EventHandler listener : listeners) {
                listener.execute(event);
            }
        }
    }

    @Override
    public <E extends Event> void callAsync(@NotNull E event) {
        if (event instanceof EventCancellable) throw new IllegalArgumentException("Cancellable event cannot be run async");
        asyncExecutor.execute(() -> call(event));
    }

    @Override
    public void register(@NotNull Object addon, @NotNull Object listener) {
        checkIfAddon(addon);

        Method[] publicMethods = listener.getClass().getMethods();
        Method[] privateMethods = listener.getClass().getDeclaredMethods();

        Set<Method> methods = new HashSet<>(publicMethods.length + privateMethods.length);
        Collections.addAll(methods, publicMethods);
        Collections.addAll(methods, privateMethods);

        for (Method method : methods) {
            EventSubscribe entry = method.getAnnotation(EventSubscribe.class);
            if (entry == null || method.isBridge() || method.isSynthetic())
                continue;

            Class<?>[] params = method.getParameterTypes();
            if (params.length < 1) continue;

            Class<?> clazz = params[0];
            if (params.length != 1 || !Event.class.isAssignableFrom(clazz)) continue;

            final Class<? extends Event> eventClass = clazz.asSubclass(Event.class);

            method.setAccessible(true);
            EventHandler<?> handler = (event) -> {
                try {
                    method.invoke(listener, event);
                } catch (Throwable e) {
                    LOGGER.warn("Failed to call an event: {}", e.getMessage());
                }
            };

            EnumMap<EventPriority, List<EventHandler<?>>> listeners = this.handlers.get(eventClass);
            if (listeners == null) {
                listeners = new EnumMap<>(EventPriority.class);
                handlers.put(eventClass, listeners);
            }

            listeners.compute(
                    entry.priority(),
                    (p, eventHandlers) -> {
                        if (eventHandlers == null) eventHandlers = new CopyOnWriteArrayList<>();
                        eventHandlers.add(handler);
                        return eventHandlers;
                    }
            );

            this.registeredListeners.compute(
                    listener,
                    (l, listenerHandlers) -> {
                        if (listenerHandlers == null) listenerHandlers = new CopyOnWriteArrayList<>();
                        listenerHandlers.add(handler);
                        return listenerHandlers;
                    }
            );
        }

        if (registeredListeners.containsKey(listener)) {
            registeredAddonListeners.compute(
                    addon,
                    (a, addonListeners) -> {
                        if (addonListeners == null) addonListeners = new CopyOnWriteArrayList<>();
                        addonListeners.add(listener);
                        return addonListeners;
                    }
            );
        }
    }

    @Override
    public <E extends Event> void register(@NotNull Object addon, Class<E> eventClass, EventPriority priority, @NotNull EventHandler<E> handler) {
        checkIfAddon(addon);

        EnumMap<EventPriority, List<EventHandler<?>>> listeners = this.handlers.get(eventClass);
        if (listeners == null) {
            listeners = new EnumMap<>(EventPriority.class);
            handlers.put(eventClass, listeners);
        }

        listeners.compute(
                priority,
                (p, eventHandlers) -> {
                    if (eventHandlers == null) eventHandlers = new CopyOnWriteArrayList<>();
                    eventHandlers.add(handler);
                    return eventHandlers;
                }
        );

        registeredAddonHandlers.compute(
                addon,
                (a, addonHandlers) -> {
                    if (addonHandlers == null) addonHandlers = new CopyOnWriteArrayList<>();
                    addonHandlers.add(handler);
                    return addonHandlers;
                }
        );
    }

    @Override
    public void unregister(@NotNull Object addon) {
        checkIfAddon(addon);

        List<EventHandler<?>> handlersToRemove = new ArrayList<>();

        List<Object> addonListeners = registeredAddonListeners.remove(addon);
        if (addonListeners != null) {
            for (Object listener : addonListeners) {
                List<EventHandler<?>> handlers = registeredListeners.remove(listener);
                if (handlers != null) handlersToRemove.addAll(handlers);
            }
        }

        List<EventHandler<?>> addonHandlers = registeredAddonHandlers.remove(addon);
        if (addonHandlers != null) handlersToRemove.addAll(addonHandlers);

        if (handlersToRemove.size() > 0) removeHandlers(handlersToRemove);
    }

    @Override
    public void unregister(@NotNull Object addon, @NotNull Object listener) {
        checkIfAddon(addon);

        List<Object> addonListeners = registeredAddonListeners.get(addon);
        addonListeners.remove(listener);
        if (addonListeners.size() == 0) registeredAddonListeners.remove(addon);

        List<EventHandler<?>> listenerHandlers = registeredListeners.remove(listener);

        if (listenerHandlers != null && listenerHandlers.size() > 0) removeHandlers(listenerHandlers);
    }

    @Override
    public void unregister(@NotNull Object addon, @NotNull EventHandler<?> handler) {
        checkIfAddon(addon);

        List<EventHandler<?>> addonHandlers = registeredAddonHandlers.get(addon);
        addonHandlers.remove(handler);
        if (addonHandlers.size() == 0) registeredAddonHandlers.remove(addon);

        removeHandlers(ImmutableList.of(handler));
    }

    private void checkIfAddon(@NotNull Object addon) {
        if (!addon.getClass().isAnnotationPresent(Addon.class) && !(addon instanceof PlasmoVoice)) {
            throw new IllegalArgumentException("object is not annotated with @Addon");
        }
    }

    private void removeHandlers(List<EventHandler<?>> handlersToRemove) {
        List<Class<? extends Event>> eventsToRemove = new ArrayList<>();

        handlers.forEach((eventClass, listeners) -> {
            List<EventPriority> listenersToRemove = new ArrayList<>();

            listeners.forEach((priority, handlers) -> {
                handlers.removeAll(handlersToRemove);
                if (handlers.size() == 0) {
                    listenersToRemove.add(priority);
                }
            });

            listenersToRemove.forEach(listeners::remove);
            if (listeners.size() == 0) {
                eventsToRemove.add(eventClass);
            }
        });

        eventsToRemove.forEach(handlers::remove);
    }
}

File: common/src/main/kotlin/su/plo/voice/event/VoiceEventBus.kt
package su.plo.voice.event

import com.google.common.collect.Maps
import su.plo.voice.BaseVoice
import su.plo.voice.api.event.Event
import su.plo.voice.api.event.EventBus
import su.plo.voice.api.event.EventCancellable
import su.plo.voice.api.event.EventHandler
import su.plo.voice.api.event.EventPriority
import su.plo.voice.api.event.EventSubscribe
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService

private val logger = BaseVoice.createLogger(VoiceEventBus::class.java.simpleName)

class VoiceEventBus(
    private val asyncExecutor: ExecutorService,
) : EventBus {
    private val handlersByListeners: MutableMap<Any, MutableList<EventHandler<*>>> = HashMap()
    private val listenersByAddons: MutableMap<Any, MutableList<Any>> = HashMap()
    private val handlersByAddons: MutableMap<Any, MutableList<EventHandler<*>>> = HashMap()

    private val handlers: MutableMap<Class<*>, HandlerList> = Maps.newConcurrentMap()

    override fun <E : Event> fire(event: E): Boolean {
        val handlerList = this.handlers[event.javaClass] ?: return true

        for (eventHandler in handlerList.handlers) {
            eventHandler.execute(event)
        }

        if (event is EventCancellable) {
            return !event.isCancelled()
        }

        return true
    }

    override fun <E : Event> fireAsync(event: E): CompletableFuture<E> {
        val future = CompletableFuture<E>()

        asyncExecutor.execute {
            fire(event)
            future.complete(event)
        }

        return future
    }

    @Synchronized
    override fun register(addon: Any, listener: Any) {
        val methods = (listener.javaClass.methods + listener.javaClass.declaredMethods).distinct()
        methods
            .filter { !it.isBridge && !it.isSynthetic }
            .filter { it.parameterCount == 1 }
            .filter { it.isAnnotationPresent(EventSubscribe::class.java) }
            .map { method ->
                val annotation = method.getAnnotation(EventSubscribe::class.java)

                method to annotation
            }
            .forEach { (method, annotation) ->
                val eventClass = method.parameterTypes[0]
                    .takeIf { Event::class.java.isAssignableFrom(it) }
                    ?.asSubclass(Event::class.java)
                    ?: return@forEach

                method.isAccessible = true

                val handler = EventHandler<Event> { event ->
                    if (annotation.ignoreCancelled &&
                        event is EventCancellable &&
                        event.isCancelled()
                    ) return@EventHandler

                    try {
                        method.invoke(listener, event)
                    } catch (e: Throwable) {
                        logger.warn("Failed to fire an event", e)
                    }
                }

                addHandler(eventClass, annotation.priority, handler)

                handlersByListeners.getOrPut(listener) { mutableListOf() }.add(handler)
            }

        if (handlersByListeners.containsKey(listener)) {
            listenersByAddons.getOrPut(addon) { mutableListOf() }.add(listener)
        }
    }

    @Synchronized
    override fun <E : Event> register(
        addon: Any,
        eventClass: Class<E>,
        priority: EventPriority,
        handler: EventHandler<E>,
    ) {
        addHandler(eventClass, priority, handler.eraseEventType())

        handlersByAddons.getOrPut(addon) { mutableListOf() }.add(handler)
    }

    @Synchronized
    override fun unregister(addon: Any) {
        val handlersToRemove = mutableListOf<EventHandler<*>>()

        listenersByAddons.remove(addon)
            ?.mapNotNull { handlersByListeners.remove(it) }
            ?.forEach { handlersToRemove.addAll(it) }

        handlersByAddons.remove(addon)
            ?.let { handlersToRemove.addAll(it) }

        if (handlersToRemove.isNotEmpty()) removeHandlers(handlersToRemove)
    }

    @Synchronized
    override fun unregister(addon: Any, listener: Any) {
        val addonListeners = listenersByAddons[addon]
        if (addonListeners != null) {
            addonListeners.remove(listener)
            if (addonListeners.isEmpty()) listenersByAddons.remove(addon)
        }

        handlersByListeners.remove(listener)
            ?.let { removeHandlers(it) }
    }

    @Synchronized
    override fun unregister(addon: Any, handler: EventHandler<*>) {
        val addonHandlers = handlersByAddons[addon] ?: return
        addonHandlers.remove(handler)
        if (addonHandlers.isEmpty()) handlersByAddons.remove(addon)

        removeHandlers(listOf(handler))
    }

    override fun hasListener(eventClass: Class<*>): Boolean =
        handlers.containsKey(eventClass)

    @Suppress("UNCHECKED_CAST")
    private fun <E : Event> EventHandler<E>.eraseEventType(): EventHandler<Event> =
        this as EventHandler<Event>

    private fun addHandler(
        eventClass: Class<out Event>,
        priority: EventPriority,
        handler: EventHandler<Event>,
    ) {
        handlers[eventClass] = handlers[eventClass]?.add(priority, handler)
            ?: HandlerList.of(priority, handler)
    }

    private fun removeHandlers(handlersToRemove: Collection<EventHandler<*>>) {
        if (handlersToRemove.isEmpty()) return

        val toRemove = handlersToRemove.toSet()

        for (eventClass in handlers.keys) {
            handlers.computeIfPresent(eventClass) { _, handlerList ->
                handlerList.remove(toRemove).takeIf { it.handlers.isNotEmpty() }
            }
        }
    }

    private class HandlerList private constructor(
        val priorities: Array<EventPriority>,
        val handlers: Array<EventHandler<Event>>,
    ) {
        fun add(priority: EventPriority, handler: EventHandler<Event>): HandlerList {
            val index = priorities.indexOfFirst { it > priority }
                .takeIf { it >= 0 }
                ?: priorities.size

            return HandlerList(
                (priorities.take(index) + priority + priorities.drop(index)).toTypedArray(),
                (handlers.take(index) + handler + handlers.drop(index)).toTypedArray(),
            )
        }

        fun remove(handlersToRemove: Set<EventHandler<*>>): HandlerList {
            val kept = handlers.indices.filter { handlers[it] !in handlersToRemove }
            if (kept.size == handlers.size) return this

            return HandlerList(
                kept.map { priorities[it] }.toTypedArray(),
                kept.map { handlers[it] }.toTypedArray(),
            )
        }

        companion object {
            fun of(priority: EventPriority, handler: EventHandler<Event>): HandlerList =
                HandlerList(arrayOf(priority), arrayOf(handler))
        }
    }
}
